tool/tsh/gcp.go

```diff
@@ -0,0 +1,443 @@
+	}
+
+	defer func() {
+		if err := app.Close(); err != nil {
+			log.WithError(err).Error("Failed to close GCP app.")
+		}
+	}()
+
+	args := cf.GCPCommandArgs
+
+	cmd := exec.Command(gcloudCLIBinaryName, args...)
+	return app.RunCommand(cmd)
+}
+
+// gcpApp is an GCP app that can start local proxies to serve GCP APIs.
+type gcpApp struct {
+	cf      *CLIConf
+	profile *client.ProfileStatus
+	app     tlsca.RouteToApp
+	secret  string
+
+	localALPNProxy    *alpnproxy.LocalProxy
+	localForwardProxy *alpnproxy.ForwardProxy
+}
+
+// newGCPApp creates a new GCP app.
+func newGCPApp(cf *CLIConf, profile *client.ProfileStatus, app tlsca.RouteToApp) (*gcpApp, error) {
+	secret, err := getGCPSecret()
+	if err != nil {
+		return nil, err
+	}
+	return &gcpApp{
+		cf:      cf,
+		profile: profile,
+		app:     app,
+		secret:  secret,
+	}, nil
+}
+
+// getGCPSecret will return fresh secret to use or read it from environment.
+func getGCPSecret() (string, error) {
+	secret := os.Getenv("TELEPORT_GCP_SECRET")
+	if secret != "" {
+		return secret, nil
+	}
+
+	return utils.CryptoRandomHex(10)
+}
+
+// StartLocalProxies sets up local proxies for serving GCP clients.
+//
+// At minimum clients should work with these variables set:
+// - HTTPS_PROXY, for routing the traffic through the proxy
+// - MSI_ENDPOINT, for informing the client about credential provider endpoint
+//
+// The request flow to remote server (i.e. GCP APIs) looks like this:
+// clients -> local forward proxy -> local ALPN proxy -> remote server
+//
+// However, with MSI_ENDPOINT variable set, clients will reach out to this address for tokens.
+// We intercept calls to https://gcp-msi.teleport.dev using alpnproxy.GCPMSIMiddleware.
+// These calls are served entirely locally, which helps the overall performance experienced by the user.
+func (a *gcpApp) StartLocalProxies() error {
+	// HTTPS proxy mode
+	if err := a.startLocalALPNProxy(""); err != nil {
+		return trace.Wrap(err)
+	}
+	if err := a.startLocalForwardProxy(a.cf.LocalProxyPort); err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+// Close makes all necessary close calls.
+func (a *gcpApp) Close() error {
+	var errs []error
+	if a.localALPNProxy != nil {
+		errs = append(errs, a.localALPNProxy.Close())
+	}
+	if a.localForwardProxy != nil {
+		errs = append(errs, a.localForwardProxy.Close())
+	}
+	return trace.NewAggregate(errs...)
+}
+
+// GetEnvVars returns required environment variables to configure the
+// clients.
+func (a *gcpApp) GetEnvVars() (map[string]string, error) {
+	envVars := map[string]string{
+		// TODO: verify secret in calls somehow
+		"CLOUDSDK_AUTH_ACCESS_TOKEN": a.secret,
+
+		// Needed for gcloud CLI to accept our certs.
+		// This isn't portable and applications may have to set different env variables,
+		// add the application cert to system root store (not recommended, ultimate fallback)
+		// or use equivalent of --insecure flag.
+		"REQUESTS_CA_BUNDLE": a.profile.AppLocalCAPath(a.app.Name),
```

Comment:
nit: add reference URL link to documentation for these env vars if possible

Reply:
Trying to find the docs for `REQUESTS_CA_BUNDLE` I realized it is non-standard env var and isn't documented anywhere. Instead, there is [`core.custom_ca_certs_file`](https://cloud.google.com/sdk/gcloud/reference/config/set#:~:text=custom_ca_certs_file) property which can be set via env variables too (`CLOUDSDK_CORE_CUSTOM_CA_CERTS_FILE`).